File: LeetCode/Problems/Python/79. Word Search.py
#DFS + Backtracking
#Time: O(n * 4^L), L is the length of word
#Space: O(n)
class Solution:
    def dfs(self, board, row, col, word, visited, idx):
        if idx == len(word):
            return True      
        directions = [(0,1),(1,0),(0,-1),(-1,0)]
        for d in directions:
            if 0 <= row + d[0] < len(board) and 0 <= col + d[1] < len(board[0]):
                new_i = row + d[0]
                new_j = col + d[1]
                if board[new_i][new_j] == word[idx] and not visited[new_i][new_j]:
                    visited[new_i][new_j] = True
                    ret = self.dfs(board, new_i, new_j, word, visited, idx+1)
                    if ret:
                        return True
                    visited[new_i][new_j] = False
        return False
                            
                            
                        
        
    def exist(self, board: List[List[str]], word: str) -> bool:
        #intuition is we can traverse the board using dfs with a visited map to perform backtracking
        #we can look at 4 neighbors
        if not board or not word:
            return False
        
        visited = [[False] * len(board[0]) for _ in board]
        for i in range(len(board)):
            for j in range(len(board[0])):
                if board[i][j] == word[0]:
                    print("in:", i,j )
                    visited[i][j] = True
                    ret = self.dfs(board, i, j, word, visited, 1)
                    if ret:
                        return True
                    else:
                        visited[i][j] = False
    
        
        

File: LeetCode/Problems/C++/79. Word Search.cpp
//Time: O(m*n*4^L)
//Space: O(L)
class Solution {
public:
    bool dfs(int i, int j, vector<vector<char>>& board, vector<vector<bool>>& visited, string& word, int idx) {
        if(idx==word.size())
            return true;
        int M = board.size();
        int N = board[0].size();
        if(i >= M || i < 0 || j >= N || j < 0) return false;
        
        if(board[i][j] == word[idx]){
            board[i][j] = '#';
            if(dfs(i+1, j, board, visited, word, idx+1)) return true;
            if(dfs(i-1, j, board, visited, word, idx+1)) return true;
            if(dfs(i, j+1, board, visited, word, idx+1)) return true;
            if(dfs(i, j-1, board, visited, word, idx+1)) return true;
            board[i][j] = word[idx];
        }
        return false;
    }
    
    bool exist(vector<vector<char>>& board, string word) {
        if(!board.size()) return false;
        
        int M = board.size();
        int N = board[0].size();
        
        vector<vector<bool>> visited(M, vector<bool>(N, false));
        
        for(int i=0;i<M;i++){
            for(int j=0;j<N;j++){
                if(board[i][j] == word[0]){
                    if(dfs(i, j, board, visited, word, 0))
                        return true;      
                }
            }
        }
        return false;
    }
};


class Solution {
public:
    bool dfs(int i, int j, vector<vector<char>>& board, vector<vector<bool>>& visited, string& word, int idx) {
        if(idx==word.size())
            return true;
        int M = board.size();
        int N = board[0].size();
        vector<vector<int>> direction = {{1,0}, {0,1}, {-1,0}, {0,-1}};
        
        for(vector<int> d: direction){
            int new_i = i+d[0];
            int new_j = j+d[1];
            if(new_i >=0 && new_i < M && new_j >=0 && new_j < N){
                if(!visited[new_i][new_j] && board[new_i][new_j] == word[idx]){
                    visited[new_i][new_j] = true;
                    if(dfs(new_i, new_j, board, visited, word, idx+1))
                        return true;
                    visited[new_i][new_j] = false;
                    
                }
            }
        }
        return false;
    }
    
    bool exist(vector<vector<char>>& board, string word) {
        if(!board.size()) return false;
        
        int M = board.size();
        int N = board[0].size();
        
        vector<vector<bool>> visited(M, vector<bool>(N, false));
        
        for(int i=0;i<M;i++){
            for(int j=0;j<N;j++){
                if(board[i][j] == word[0]){
                    visited[i][j] = true;
                    if(dfs(i, j, board, visited, word, 1))
                        return true;
                    visited[i][j] = false;
                }
            }
        }
        return false;
    }
};